iceoryx_binding_c/source/c_wait_set.cpp

```diff
@@ -187,17 +215,104 @@ void iox_ws_detach_subscriber_event(iox_ws_t const self,
+                                             const iox_ClientEvent clientEvent,
+                                             const uint64_t eventId,
+                                             void (*callback)(iox_client_t))
+{
+    iox::cxx::Expects(self != nullptr);
+    iox::cxx::Expects(client != nullptr);
+
+    auto result = self->attachEvent(*client, c2cpp::clientEvent(clientEvent), eventId, {callback, nullptr});
+    return (result.has_error()) ? cpp2c::waitSetResult(result.get_error()) : iox_WaitSetResult::WaitSetResult_SUCCESS;
+}
+
+iox_WaitSetResult iox_ws_attach_client_event_with_context_data(iox_ws_t const self,
+                                                               iox_client_t const client,
+                                                               const ENUM iox_ClientEvent clientEvent,
+                                                               const uint64_t eventId,
+                                                               void (*callback)(iox_client_t, void*),
+                                                               void* const contextData)
+{
+    iox::cxx::Expects(self != nullptr);
+    iox::cxx::Expects(client != nullptr);
+
+    NotificationCallback<std::remove_pointer_t<iox_client_t>, void> notificationCallback;
+    notificationCallback.m_callback = callback;
+    notificationCallback.m_contextData = contextData;
+
+    auto result = self->attachEvent(*client, c2cpp::clientEvent(clientEvent), eventId, notificationCallback);
+    return (result.has_error()) ? cpp2c::waitSetResult(result.get_error()) : iox_WaitSetResult::WaitSetResult_SUCCESS;
+}
+
+iox_WaitSetResult iox_ws_attach_client_state(const iox_ws_t self,
+                                             const iox_client_t client,
+                                             const iox_ClientState clientState,
+                                             const uint64_t eventId,
+                                             void (*callback)(iox_client_t))
+{
+    iox::cxx::Expects(self != nullptr);
+    iox::cxx::Expects(client != nullptr);
+
+    auto result = self->attachState(*client, c2cpp::clientState(clientState), eventId, {callback, nullptr});
+    return (result.has_error()) ? cpp2c::waitSetResult(result.get_error()) : iox_WaitSetResult::WaitSetResult_SUCCESS;
+}
+
+iox_WaitSetResult iox_ws_attach_client_state_with_context_data(iox_ws_t const self,
+                                                               iox_client_t const client,
+                                                               const ENUM iox_ClientState clientState,
+                                                               const uint64_t eventId,
+                                                               void (*callback)(iox_client_t, void*),
+                                                               void* const contextData)
+{
+    iox::cxx::Expects(self != nullptr);
+    iox::cxx::Expects(client != nullptr);
```

Comment:
same here, nullptr check for `contextData`?

Reply:
Nope, since `nullptr` is a valid value for `contextData`. This pointer is provided as argument to a user defined function and when the user can handle a `nullptr` we have no problem. It is the responsibility of the user.